src/Google.Logging.Log4Net/LogUploader.cs

```diff
@@ -0,0 +1,154 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Logging.V2;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Logging.Log4Net
+{
+    internal class LogUploader
+    {
+        private static readonly Dictionary<string, string> s_emptyLabels = new Dictionary<string, string>();
+
+        public LogUploader(
+            LoggingServiceV2Client client,
+            IScheduler scheduler,
+            IClock clock,
+            ILogQueue logQ,
+            LogEntry logsLostWarningTemplate,
+            int maxUploadBatchSize,
+            BackoffSettings serverErrorBackoffSettings)
+        {
+            _client = client;
+            _scheduler = scheduler;
+            _clock = clock;
+            _logQ = logQ;
+            _logsLostWarningTemplate = logsLostWarningTemplate;
+            _maxUploadBatchSize = maxUploadBatchSize;
+            _serverErrorBackoffSettings = serverErrorBackoffSettings;
+            _uploaderTask = Task.Run(RunUploader);
+        }
+
+        private readonly object _lockObj = new object();
+        private readonly LoggingServiceV2Client _client;
+        private readonly IScheduler _scheduler;
+        private readonly IClock _clock;
+        private readonly ILogQueue _logQ;
+        private readonly LogEntry _logsLostWarningTemplate;
+        private readonly int _maxUploadBatchSize;
+        private readonly BackoffSettings _serverErrorBackoffSettings;
+
+        private readonly Task _uploaderTask;
+
+        private readonly AsyncAutoResetEvent _uploadReadyEvent = new AsyncAutoResetEvent();
+        private readonly AsyncAutoResetEvent _emptyEvent = new AsyncAutoResetEvent();
+
+        private DateTimeRange _logEntriesLost = null;
+
+        public void TriggerUpload(DateTimeRange logEntriesLost)
+        {
+            lock (_lockObj)
+            {
+                _logEntriesLost = _logEntriesLost != null ? _logEntriesLost.Merge(logEntriesLost) : logEntriesLost;
+                _uploadReadyEvent.Set();
+            }
+        }
+
+        private LogEntry MakeLostEntry(DateTimeRange lostRange)
+        {
+            var lostEntry = _logsLostWarningTemplate.Clone();
+            lostEntry.Timestamp = _clock.GetCurrentDateTimeUtc().ToTimestamp();
+            lostEntry.TextPayload = string.Format(lostEntry.TextPayload,
+                lostRange.From.ToString(GoogleStackdriverAppender.s_lostDateTimeFmt),
+                lostRange.To.ToString(GoogleStackdriverAppender.s_lostDateTimeFmt));
+            return lostEntry;
+        }
+
+        private async Task RunUploader()
```

Comment:
At the moment yes.
But I've just thought that there might be a way to programmatically shutdown log4net. If there is, then this loop should probably exit after all log entries have been uploaded.
I'll put in a TODO to check.